src/main/scala/tech/sourced/gemini/Hash.scala

```diff
@@ -0,0 +1,171 @@
+package tech.sourced.gemini
+
+import gopkg.in.bblfsh.sdk.v1.uast.generated.Node
+import org.apache.spark.rdd.RDD
+import org.apache.spark.sql.{DataFrame, SparkSession}
+import org.apache.spark.sql.cassandra._
+import org.apache.spark.sql.functions._
+import org.slf4j.{Logger => Slf4jLogger}
+import tech.sourced.engine._
+import tech.sourced.featurext.SparkFEClient
+import tech.sourced.featurext.generated.service.Feature
+
+case class RDDFeatureKey(token: String, doc: String)
+case class RDDFeature(key: RDDFeatureKey, weight: Long)
+case class RDDHash(doc: String, wmh: Array[Array[Long]])
+case class HashResult(files: DataFrame, hashes: RDD[RDDHash], docFreq: OrderedDocFreq)
+
+/**
+  * Implements hashing repositories using Apache Spark
+  *
+  * @param session spark session
+  * @param log
+  */
+class Hash(session: SparkSession, log: Slf4jLogger) {
+  import session.implicits._
+
+  /**
+    * Calculates hashes and docFreq
+    *
+    * @param repos DataFrame with engine.getRepositories schema
+    */
+  def forRepos(repos: DataFrame): HashResult = {
+    val files = filesForRepos(repos).cache()
+    val uasts = extractUast(files).cache()
+    val features = extractFeatures(uasts).cache()
+    val docFreq = makeDocFreq(uasts, features)
+    val hashes = hashFeatures(docFreq, features)
+
+    HashResult(files, hashes, docFreq)
+  }
+
+  /**
+    * Save hashing results to database
+    *
+    * @param hashResult
+    * @param keyspace
+    * @param tables
+    */
+  def save(hashResult: HashResult, keyspace: String, tables: Tables): Unit ={
+    saveMeta(hashResult.files, keyspace, tables)
+    saveDocFreq(hashResult.docFreq)
+    saveHashes(hashResult.hashes, keyspace, tables)
+  }
+
+  protected def filesForRepos(repos: DataFrame): DataFrame =
+    repos
+      .getHEAD
+      .getCommits
+      .getTreeEntries
+      .getBlobs
+      .filter('is_binary === false)
+
+  protected def extractUast(files: DataFrame): DataFrame = {
+    // TODO(max): get languages from bblfsh directly as soon as
+    // https://github.com/bblfsh/client-scala/issues/68 resolved
+    val langs = Array("Java", "Python", "Go", "JavaScript", "TypeScript", "Ruby", "Bash", "Php")
```

Comment:
Prefer `Seq` by default, unless you actually want a specific implementation and use `Array`, `List`, etc.